File: 0216-combination-sum-iii/0216-combination-sum-iii.cpp
class Solution {
public:
    vector<vector<int>> ans;
    void helper(int index, int n, int k, vector<int>& arr, vector<int>& curr) {
        if (n == 0 && curr.size() == k){
            ans.push_back(curr);
            return;
        }
        if (index == arr.size()) return;
        for (int i = index; i < arr.size(); i++) {
            if (i > index && arr[i] == arr[i-1]) continue;
            if (!curr.empty() || n >= arr[i]) {
                curr.push_back(arr[i]);
                helper(i+1, n-arr[i], k, arr, curr);
                curr.pop_back();
            }
        }
    }
    vector<vector<int>> combinationSum3(int k, int n) {
        vector<int> arr(9);
        for (int i = 0; i < 9; i++) {
            arr[i] = i + 1;
        }
        vector<int> curr;
        helper(0, n, k, arr, curr);
        return ans;
    }
};

File: 0216-combination-sum-iii/0216-combination-sum-iii.java
class Solution {
    List<List<Integer>> ans = new ArrayList<>();
    public void helper(int index, int k, int n, int[] arr, List<Integer> curr) {
        if (n == 0 && curr.size() == k) {
            ans.add(new ArrayList<>(curr));
            return;
        }
        if (index == arr.length) return;
        for (int i = index; i < arr.length; i++) {
            if (i > index && arr[i] == arr[i-1]) continue;
            if (!curr.isEmpty() || n >= arr[i]) {
                curr.add(arr[i]);
                helper(i+1, k, n-arr[i], arr, curr);
                curr.remove(curr.size()-1);
            }
        }
    }
    public List<List<Integer>> combinationSum3(int k, int n) {
        int[] arr = new int[9];
        for (int i = 0; i < 9; i++) {
            arr[i] = i + 1;
        }
        helper(0, k, n, arr, new ArrayList<>());
        return ans;
    }
}